lib/Tests.js

```diff
@@ -341,41 +302,113 @@ TestsModule.prototype.search = async function(txt,match,limit,skip)
 
-  // pad with zeroes and "F"s.
-  var qlow = q.padEnd(24,'0');
-  var qhigh = q.padEnd(24,'F');
-  // logger.info(qlow,qhigh);
-  var bitlow = ObjectID(qlow);
-  var bithigh =  ObjectID(qhigh);
-  
-  var query = {_id:{$gte: bitlow, $lte: bithigh}};
-  if(formId) query.formId = formId;
+    // supplied by API:
+    //{}
+    // _id: <ObjectId>,    // Auto-set by Mongo, Includes insertion timestamp redundantly.
+    // recordType: <string>   //  "test" or "job" respectively
+    // insertion: <insertion block>
 
-  // logger.info(qlow,qhigh,bitlow,bithigh);
-  logger.info(query,"QfindTestIdStartsWith");
- 
-  var matches = await db.collection("tests")
-    .find(query)  // binary ObjectIds only
-    .project({"_id":1, 'componentUuid':1, 'formId':1})
-    .limit(max)
-    .toArray();
-  // if(matches.length>=max) return new Error("Too many entries"); // too many!
-  for(var m of matches) {
-    try {  m.componentUuid = MUUID.from(m.componentUuid).toString(); } catch {};
-  }
-  return matches;
+    // Returns the ObjectID of the saved record.
+    //
+    // req must contain user and ip objects.
 
-}
+    if(!input) throw new Error("saveTestData: no input data provided");
+    if(!input.formId) throw new Error("saveTestData: formId not defined");
+    if(!input.data) throw new Error("saveTestData: no data given");
 
+    var list = await Forms.list(this._formCollection);
+    var formInfo = list[input.formId];
+    // logger.info("formId",input.formId);
+    // logger.info("formInfo",formInfo);
+    // logger.info("list",list);
+    if(!formInfo) {
+        // This form doesn't actually exit.
+        throw new Error("saveTestData: '"+input.formId+"' is not a known formId");
+        // formInfo = {};
+    }
+
+    var record = {...input};
+
+    if(!input.formObjectId) {
+      record.formObjectId = formInfo.formObjectId;
+    } else {
+      record.formObjectId = new ObjectID(input.formObjectId);
+      if(!ObjectID.isValid(record.formObjectId) ) throw new Error("saveTestData: formObjectId not a correct formId");
+    }
+    if(!input.formName) {
+      record.formName = formInfo.formName;
+    }
+
+    // Ensure componentID in binary form.
+    if(this._type === "test") {
+        if(!input.componentUuid) throw new Error("Tests::save() No componentUuid provided.");
+        record.componentUuid = MUUID.from(input.componentUuid);
+    } 
+    record.recordType = this._type; // required
+    // logger.info('req.user',req.user);
+    record.insertion = commonSchema.insertion(req);
+
+    // metadata.
+    record.state = record.state || 'submitted'; // ensure 'state' is set.
+    // logger.info("Tests::save()",record.formId,record.state);
+
+    if(!permissions.hasPermission(req,this._collection+':submit'))  throw new Error("You don't have permission to edit forms.");
+    
+    var result;
+    var draft_id = input._id;
+
+    // If it's a draft and already has a record number, replace it.
+    if(input.state == "draft" && draft_id) {
```

Comment:
Just a nitpick, but we should always use type-constrained comparisons:  `===`.

Reply:
Changed.  I'll clean up that 'save' function as a whole with the rest of the API-related functions sometime later.